### PR TITLE
[SMTChecker] User timeout option

### LINKED ISSUE #9876: [SMTChecker] User timeout option
Currently the SMTChecker uses a resource limit per SMT query. `rlimit` was chosen instead of the usual timeout to improve reproducibility in the SMTChecker tests. However, that's not very helpful for users.

We should allow a user configurable timeout for SMT queries.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/9876

### CODE REVIEW

File: docs/using-the-compiler.rst [hunk 1]
```diff
@@ -346,7 +346,9 @@ Input Description
       "modelCheckerSettings":
       {
         // Choose which model checker engine to use: all (default), bmc, chc, none.
-        "engine": "chc"
+        "engine": "chc",
+        // Timeout for each SMT query in milliseconds.
```
Comment (axic) [inline docs/using-the-compiler.rst:350]: Shouldn't there be a comment here what is the default if this is omitted?

File: libsmtutil/Z3CHCInterface.cpp [hunk 2]
```diff
@@ -97,7 +102,8 @@ pair<CheckResult, CHCSolverInterface::CexGraph> Z3CHCInterface::query(Expression
 	}
 	catch (z3::exception const& _err)
 	{
-		if (_err.msg() == string("max. resource limit exceeded"))
+		set<string> msgs{"max. resource limit exceeded", "canceled"};
```
Comment (axic) [inline libsmtutil/Z3CHCInterface.cpp]: This tricked me again: https://www.merriam-webster.com/words-at-play/canceled-or-cancelled

Comment (axic) [inline libsmtutil/Z3CHCInterface.cpp]:
```suggestion
		set<string> msgs{
		    // Resource limit exhausted.
		    "max. resource limit exceeded",
		    // User submitted timeout exhausted.
		    "canceled"
		};
```

File: libsmtutil/Z3Interface.cpp [hunk 3]
```diff
@@ -104,9 +109,13 @@ pair<CheckResult, vector<string>> Z3Interface::check(vector<Expression> const& _
 				values.push_back(util::toString(m.eval(toZ3Expr(e))));
 		}
 	}
-	catch (z3::exception const&)
+	catch (z3::exception const& _err)
 	{
-		result = CheckResult::ERROR;
+		set<string> msgs{"max. resource limit exceeded", "canceled"};
```
Comment (axic) [inline libsmtutil/Z3Interface.cpp]: If you decide to take the previous suggestion, please take it here too.

File: libsolidity/interface/StandardCompiler.cpp [hunk 4]
```diff
@@ -885,7 +885,14 @@ std::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompiler:
...
 		if (!engine)
 			return formatFatalError("JSONError", "Invalid model checker engine requested.");
-		ret.modelCheckerEngine = *engine;
+		ret.modelCheckerSettings.engine = *engine;
+	}
+
+	if (modelCheckerSettings.isMember("timeout"))
+	{
+		if (!modelCheckerSettings["timeout"].isUInt())
+			return formatFatalError("JSONError", "modelCheckerSettings.timeout must be an unsigned integer.");
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp:894]: The error message for the model checker is a tad differently styled. See above.

File: solc/CommandLineInterface.cpp [hunk 5]
```diff
@@ -1002,6 +1004,11 @@ General Information)").c_str(),
 			po::value<string>()->value_name("all,bmc,chc,none")->default_value("all"),
 			"Select model checker engine."
 		)
+		(
+			g_strModelCheckerTimeout.c_str(),
+			po::value<unsigned>()->value_name("ms"),
+			"Set model checker timeout per query in milliseconds. 0 means no timeout."
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: Well, 0 means forcefully no timeout, while omitting the option means a resource limit. So maybe mentioning these in both the standard json part and here would be nice.

Comment (leonardoalt) [reply solc/CommandLineInterface.cpp]: Will do

File: libsmtutil/Z3CHCInterface.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsmtutil/Z3CHCInterface.cpp]: Just copied directly from the solver output

File: docs/using-the-compiler.rst (hunk 1 above)
Comment (leonardoalt) [reply docs/using-the-compiler.rst:350]: Ok!

**Comment by leonardoalt** [comment]
Updated

**Review by axic — APPROVED** [review]
Haven't done a nitpicking level review, but the idea, the documentation, the error messages all look nice.

**Comment by axic** [comment]
As a comment it is a concerning how much duplication exists between the different interfaces/solvers.

**Comment by leonardoalt** [comment]
@axic Yea this came up also when I first wrote the CHC interface and solvers. They do have common stuff, but it's hard to put them  together, so it's easier and clearer to keep the duplication.

**Pull request merged**